Mathlib/Topology/Order/HullKernel.lean

```diff
@@ -0,0 +1,250 @@
+  constructor <;> intro ha
+  · obtain ⟨a, ha'⟩ :=  ha
+    use {a}
+    rw [← (Function.Injective.preimage_image Subtype.val_injective R), ← ha']
+    simp only [finite_singleton, upperClosure_singleton, UpperSet.coe_Ici, image_val_compl,
+      Subtype.image_preimage_coe, diff_self_inter, preimage_diff, Subtype.coe_preimage_self,
+      true_and]
+    exact compl_eq_univ_diff (Subtype.val ⁻¹' Ici a)
+  · obtain ⟨F, hF⟩ := ha
+    lift F to Finset α using hF.1
+    use Finset.inf F id
+    rw [← (preimage_upperClosure_finset hT), ← hF.2]
+    rfl
+
+end PrimitiveSpectrum
+
+end SemilatticeInf
+
+section PrimativeSpectrum
+
+variable [CompleteLattice α]
+
+variable {T : Set α}
+
+namespace PrimitiveSpectrum
+
+lemma sInter_hull (S : Set α) : ⋂₀ { hull T a | a ∈ S } = hull T (sSup S) := by
+  ext x : 1
+  simp_all only [mem_sInter, mem_setOf_eq, forall_exists_index, and_imp, forall_apply_eq_imp_iff₂,
+    mem_preimage, mem_Ici, sSup_le_iff]
+
+/- When `α` is complete, the relative basis for the lower topology is also closed under arbitrary
+unions. -/
+lemma sUnion_hull_compl (S : Set α) : ⋃₀ { (hull T a)ᶜ | a ∈ S } = (hull T (sSup S))ᶜ := by
+  simp [sUnion_eq_compl_sInter_compl, ← sInter_hull, compl_sInter]
+
+/- When `α` is complete, a set is Lower topology relative-open if and only if it is of the form
+`(hull T a)ᶜ` for some `a` in `α`.-/
+lemma isOpen_iff [TopologicalSpace α] [IsLower α] [DecidableEq α] (hT : ∀ p ∈ T, InfPrime p)
+    (S : Set T) : IsOpen S ↔ ∃ (a : α), S = (hull T a)ᶜ := by
+  constructor <;> intro h
+  · let R := {a : α | (hull T a)ᶜ ⊆ S}
+    use sSup R
+    rw [IsTopologicalBasis.open_eq_sUnion' (relativeLowerIsTopologicalBasis hT) h]
+    aesop
+  · obtain ⟨a, ha⟩ := h
+    exact ⟨(Ici a)ᶜ, ⟨isOpen_compl_iff.mpr isClosed_Ici, ha.symm⟩⟩
+
+/- When `α` is complete, a set is closed in the relative lower topology if and only if it is of the
+form `hull T a` for some `a` in `α`.-/
+lemma isClosed_iff [TopologicalSpace α] [IsLower α] [DecidableEq α] (hT : ∀ p ∈ T, InfPrime p)
+    (S : Set T) : IsClosed S ↔ ∃ (a : α), S = hull T a := by
+  simp only [← isOpen_compl_iff, (isOpen_iff hT), preimage_compl, compl_inj_iff]
+
+/-- For a subset `S` of `T`, `kernel S` is the infimum of `S` (considered as a set of `α`) -/
+abbrev kernel (S : Set T) := sInf (Subtype.val '' S)
+
+/- The pair of maps `S ↦ kernel S` and `a ↦ hull a` form an antitone Galois connection betwen the
+subsets of `T` and `α`. -/
+open OrderDual in
+theorem gc : GaloisConnection (α := Set T) (β := αᵒᵈ)
+    (fun S => toDual (kernel S)) (fun a => hull T (ofDual a)) := fun S a => by
+  simp only [toDual_sInf, sSup_le_iff, mem_preimage, mem_image, Subtype.exists, exists_and_right,
+      exists_eq_right, ← ofDual_le_ofDual, forall_exists_index, OrderDual.forall, ofDual_toDual]
+  exact ⟨fun h b hbS => h _ (Subtype.coe_prop b) hbS, fun h b _ hbS => h hbS⟩
+
+lemma gc_closureOperator (S : Set T) : gc.closureOperator S = hull T (kernel S) := by
```

Comment:
I would make this a simp lemma

Reply:
@YaelDillies we get:
```
/- The `simpNF` linter reports:
SOME SIMP LEMMAS ARE NOT IN SIMP-NORMAL FORM.
see note [simp-normal form] for tips how to debug this.
https://leanprover-community.github.io/mathlib_docs/notes.html#simp-normal%20form -/
-- Mathlib.Topology.Order.HullKernel
Error: /home/lean/actions-runner/_work/mathlib4/mathlib4/pr-branch/Mathlib/Topology/Order/HullKernel.lean:181:1: error: @PrimitiveSpectrum.gc_closureOperator Left-hand side simplifies from
  ⋯.closureOperator S
to
  PrimitiveSpectrum.hull T (sInf (⇑OrderDual.toDual ⁻¹' (⇑OrderDual.ofDual ⁻¹' (Subtype.val '' S))))
using
  simp only [*, @toDual_sInf, @GaloisConnection.closureOperator_apply, @ofDual_sSup]
Try to change the left-hand side to the simplified term!
```

Reply:
Swapping it around works though - is that what you meant?
```
@[simp]
lemma gc_closureOperator (S : Set T) : hull T (kernel S) = gc.closureOperator S := ...
```